Sources/RxAlamofire.swift

```diff
@@ -704,6 +704,58 @@ extension Reactive where Base: SessionManager {
         }
     }
 
+    // MARK: Multipart upload
+
+    /**
+     Returns an observable of an `UploadRequest` using the `url`, `method` and `headers`,
+     and a body that contains `multipart/form-data` content encoded from the input data,
+     provided by the `multipartFormData` parameter.
+
+     It is important to understand the memory implications of uploading `MultipartFormData`. If the cummulative
+     payload is small, encoding the data in-memory and directly uploading to a server is the by far the most
+     efficient approach. However, if the payload is too large, encoding the data in-memory could cause your app to
+     be terminated. Larger payloads must first be written to disk using input and output streams to keep the memory
+     footprint low, then the data can be uploaded as a stream from the resulting file. Streaming from disk MUST be
+     used for larger payloads such as video content.
+
+     The `encodingMemoryThreshold` parameter allows Alamofire to automatically determine whether to encode in-memory
+     or stream from disk. If the content length of the `MultipartFormData` is below the `encodingMemoryThreshold`,
+     encoding takes place in-memory. If the content length exceeds the threshold, the data is streamed to disk
+     during the encoding process. Then the result is uploaded as data or as a stream depending on which encoding
+     technique was used.
+
+     - parameter multipartFormData:       The closure used to append body parts to the `MultipartFormData`.
+     - parameter encodingMemoryThreshold: The encoding memory threshold in bytes.
+                                          `multipartFormDataEncodingMemoryThreshold` by default.
+     - parameter url:                     The URL.
+     - parameter method:                  The HTTP method. `.post` by default.
+     - parameter headers:                 The HTTP headers. `nil` by default.
+    */
+    public func upload(multipartFormData data: @escaping (MultipartFormData) -> Void,
+                       usingThreshold encodingMemoryThreshold: UInt64 = SessionManager.multipartFormDataEncodingMemoryThreshold,
+                       to url: URLConvertible,
+                       method: HTTPMethod = .post,
+                       headers: HTTPHeaders = [:]) -> Observable<UploadRequest> {
+        let manager = base
+        return Observable.create { observer in
```

Comment:
Hmm, is it possible to use [this `request()`](https://github.com/vzsg/RxAlamofire/blob/04ee1d13557fddf9652767476994d461e32cce77/Sources/RxAlamofire.swift#L415-L442) function that `upload()` above uses as well? It handles many cases like cancelling the request when disposed etc.

Reply:
I'll see what I can do.
TBH I don't have the slightest idea what `responseWith` and this `request` dance does.

Reply:
Well, it seems `request` expects a function that synchronously returns an RxAlamofireRequest – which is not applicable here, as `upload(multipartFormData:to:)` is asynchronous... 🤔

Reply:
@vzsg Hmm. We can get the `UploadRequest` by adding [`encodingCompletion`](https://github.com/Alamofire/Alamofire/blob/c8a4d3a910bcfd53b9b855463495b8200ca80fd6/Source/SessionManager.swift#L613) to the `upload(multipart:)` function. And because `UploadRequest` is also a `DataRequest`, it should be also be a `RxAlamofireRequest`.

There might be a problem with returning the request in a closure because you would probably need to not return a closure, but to execute another closure with the request. But it should be doable. This way we will get unified code and all the checks needed for a request handling.

Reply:
The encodingCompletion is [here](https://github.com/vzsg/RxAlamofire/blob/04ee1d13557fddf9652767476994d461e32cce77/Sources/RxAlamofire.swift#L746-L752), that's how the extension works in the first place. In the success case, the request is indeed available, but I'm not entirely following what you mean by the unified handling.

Could you help out with a commit or some guidance please?